In [21]:
import tensorflow as tf
import cv2
import numpy as np
import os

In [2]:
import numpy as np

In [4]:
img=np.zeros((3,3), dtype=np.uint8)

In [5]:
img

array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]], dtype=uint8)

In [6]:
img=cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)

In [7]:
img


array([[[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]]], dtype=uint8)

### KNOWING THE DIMENSIONS OF AN IMAGE

In [10]:
img=np.zeros((5, 3), dtype=np.uint8)
img.shape

(5, 3)

### Converting the image from one format to another

In [12]:
myImg=cv2.imread("mario.png")

In [13]:
myImg.shape

(1854, 1473, 3)

### Here in the above image we can see that the height of the image is in the rows i.e, 1854, while the width of the image is in the columns ie, 1473

In [15]:
cv2.imwrite("mario.jpg", myImg)

True

In [19]:
myImg_greyscale=cv2.imread("baby.jpg", cv2.IMREAD_GRAYSCALE)

In [20]:
cv2.imwrite("baby_greyscale.jpg", myImg_greyscale)

True

### Construction of a random image using random byte array

In [22]:
# Make an array of 120,000 random bytes.
# we can also use numpy.random.randint(0, 256, 120000) in place of below code
randomByteArray=bytearray(os.urandom(120000))
flatNumpyArray=np.array(randomByteArray)

In [23]:
flatNumpyArray

array([231,   6, 204, ...,  53,  44,  41], dtype=uint8)

In [25]:
# Convert the array to make a 400x300 grayscale image.
grayImage=flatNumpyArray.reshape(300, 400)
cv2.imwrite("my_random_img_greyscale.png", grayImage)

True

In [26]:
# Convert the array to make a 400x100 color image.
colorImage=flatNumpyArray.reshape(100, 400, 3)
cv2.imwrite("my_random_img_color.png", colorImage)

True

## Image Manipulations

### Say you want to manipulate a pixel at coordinates (0, 0) in a BGR image and turn it into a white pixel:

In [34]:
img=cv2.imread("baby.jpg")
img.shape

img[0, 0]=[255, 255, 255]

cv2.imwrite("baby_modified.jpg", img)
# you will see a white dot in the top-left corner of the image

True

### change the value of the blue channel at (150, 120) from its current value to an arbitrary 255:

In [35]:
img=cv2.imread("baby.jpg")
img.itemset((150, 120, 0), 255) # Sets the value of a pixel's blue channel

print(img.item(150, 120, 0)) # Prints the value of a pixel's blue channel


255


### an example of using array slicing to manipulate color channels. Setting all G (green) values of an image to 0 is as simple as the following code:

In [38]:
img=cv2.imread("baby.jpg")


img[:, :, 1] = 0
